hphp/runtime/vm/jit/relocation-arm.cpp

```diff
@@ -0,0 +1,993 @@
+      jt.first = adjusted;
+    }
+  }
+
+  if (!meta.bcMap.empty()) {
+    /*
+     * Most of the time we want to adjust to a corresponding "before" address
+     * with the exception of the start of the range where "before" can point to
+     * the end of a previous range.
+     */
+    auto const aStart = meta.bcMap[0].aStart;
+    auto const acoldStart = meta.bcMap[0].acoldStart;
+    auto const afrozenStart = meta.bcMap[0].afrozenStart;
+    auto adjustAddress = [&](TCA& address, TCA blockStart) {
+      if (TCA adjusted = (address == blockStart
+                            ? rel.adjustedAddressAfter(blockStart)
+                            : rel.adjustedAddressBefore(address))) {
+        address = adjusted;
+      }
+    };
+    for (auto& tbc : meta.bcMap) {
+      adjustAddress(tbc.aStart, aStart);
+      adjustAddress(tbc.acoldStart, acoldStart);
+      adjustAddress(tbc.afrozenStart, afrozenStart);
+    }
+  }
+
+  decltype(meta.addressImmediates) updatedAI;
+  for (auto addrImm : meta.addressImmediates) {
+    if (TCA adjusted = rel.adjustedAddressAfter(addrImm)) {
+      updatedAI.insert(adjusted);
+    } else if (TCA odd = rel.adjustedAddressAfter((TCA)~uintptr_t(addrImm))) {
+      // just for cgLdObjMethod
+      updatedAI.insert((TCA)~uintptr_t(odd));
+    } else {
+      updatedAI.insert(addrImm);
+    }
+  }
+  updatedAI.swap(meta.addressImmediates);
+
+  decltype(meta.alignments) updatedAF;
+  for (auto af : meta.alignments) {
+    if (TCA adjusted = rel.adjustedAddressAfter(af.first)) {
+      updatedAF.emplace(adjusted, af.second);
+    } else {
+      updatedAF.emplace(af);
+    }
+  }
+  updatedAF.swap(meta.alignments);
+
+  for (auto& af : meta.reusedStubs) {
+    if (TCA adjusted = rel.adjustedAddressAfter(af)) {
+      af = adjusted;
+    }
+  }
+
+  decltype(meta.smashableLocations) updatedSL;
+  for (auto sl : meta.smashableLocations) {
+    if (TCA adjusted = rel.adjustedAddressAfter(sl)) {
+      updatedSL.insert(adjusted);
+    } else {
+      updatedSL.insert(sl);
+    }
+  }
+  updatedSL.swap(meta.smashableLocations);
+
+  if (asmInfo) {
+    assert(asmInfo->validate());
+    fixupRanges(asmInfo, AreaIndex::Main, rel);
+    fixupRanges(asmInfo, AreaIndex::Cold, rel);
+    fixupRanges(asmInfo, AreaIndex::Frozen, rel);
+    assert(asmInfo->validate());
+  }
+}
+
+void findFixups(TCA start, TCA end, CGMeta& meta) {
+  while (start != end) {
+    assert(start < end);
+    Instruction* instr = Instruction::Cast(start);
+    start = reinterpret_cast<TCA>(instr->NextInstruction());
+
+    // If instruction is a call
+    if ((instr->Mask(UnconditionalBranchMask) == BL) ||
+        (instr->Mask(UnconditionalBranchToRegisterMask) == BLR)) {
+      if (auto fixup = FixupMap::findFixup(start)) {
+        meta.fixups.emplace_back(start, *fixup);
+      }
+      if (auto ct = getCatchTrace(start)) {
+        meta.catches.emplace_back(start, *ct);
+      }
+    }
+  }
+}
```

Comment:
Looks like this could be largely shared between the three architectures also, with maybe just a small handful of arch-specific metadata adjustments.  Can you nix this from the individual `relocation-xxx.h` APIs, and just have one shared implementation in `relocation.cpp`?

It should be easy to diff the existing individual implementations; they're largely identical—I think it's just that x64 does literals and code pointers, ARM does smashables, and PPC does neither.